cma/warm_starting_cma.py

```diff
@@ -0,0 +1,47 @@
+"""
+Optuna example using warm starting CMA-ES.
+
+In this example, we first optimize a *biased* sphere function as a source task.
+With the result of optimization, we optimize a sphere function as a target task
+by using warm starting CMA-ES.
+
+"""
+
+import optuna
+
+
+# 2-dimensional box-constrained sphere function
+# The optimum is (x, y) = (0, 0)
+def sphere(trial):
+    x = trial.suggest_float("x", -15, 15)
+    y = trial.suggest_float("y", -15, 15)
+    return x ** 2 + y ** 2
+
+
+# 2-dimensional box-constrained sphere function
+# This function is *biased*; the optimum is (x, y) = (1, 1)
+def biased_sphere(trial):
+    x = trial.suggest_float("x", -15, 15)
+    y = trial.suggest_float("y", -15, 15)
+    return (x - 1) ** 2 + (y - 1) ** 2
+
+
+if __name__ == "__main__":
+
+    # Perform optimization on a source task
+    cma = optuna.samplers.CmaEsSampler()
+    source_study = optuna.create_study(sampler=cma)
+    source_study.optimize(biased_sphere, n_trials=50)
+    print(
+        f"Best value on the source task: {source_study.best_value},"
+        f" (params: {source_study.best_params}\n"
+    )
+
+    # Perform optimization on a target task by warm starting CMA-ES
+    ws_cma = optuna.samplers.CmaEsSampler(source_trials=source_study.trials)
+    target_study = optuna.create_study(sampler=ws_cma)
+    target_study.optimize(sphere, n_trials=50)
```

Comment:
Just curious but is this 50 based on some reference? Reducing it to e.g. 10 would make the difference more clear (and hence more pedagogic?) but maybe it's too much.

Reply:
I do not have a special intention for the number 50, but I often use it.
The warm starting CMA-ES is, roughly speaking, constructed by the two components, (1) initialization and (2) CMA-ES.
When we use a quite small number of evaluations, it only uses the part of initialization, not CMA-ES (because it cannot be adapted from observations).
So if we want to check the behavior of warm starting CMA-ES including CMA-ES, we need a moderate number of evaluations.
In fact, in [the original paper of warm starting CMA-ES](https://arxiv.org/abs/2012.06932), most of the experiments are performed on the condition that the number of evaluations is 50.

Having said that, only the initialization is still effective. *WS only* (Fig. 3 and 5 in the paper) corresponds to this situation, and it has achieved satisfactory performance.

Reply:
Thanks for the detailed explanation. Sounds good to keep the current trial count.